llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp

```diff
@@ -2335,24 +2479,41 @@ class BoUpSLP {
                                  : cast<Instruction>(VL[0])->getNumOperands();
       OpsVec.resize(NumOperands);
       unsigned NumLanes = VL.size();
-      for (unsigned OpIdx = 0; OpIdx != NumOperands; ++OpIdx) {
+      InstructionsState S = getSameOpcode(VL, TLI);
```

Comment:
Suggest to pass it to VLOperands constructor instead and use it instead of regenerating.

Reply:
Should we only use `VLOperands`? Right now the code is like this
```
      // Sort operands of the instructions so that each side is more likely to
      // have the same opcode.
      if (isa<BinaryOperator>(VL0) && isCommutative(VL0)) {
        ValueList Left, Right;
        reorderInputsAccordingToOpcode(VL, Left, Right, *this);
        TE->setOperand(0, Left);
        TE->setOperand(1, Right);
        buildTree_rec(Left, Depth + 1, {TE, 0});
        buildTree_rec(Right, Depth + 1, {TE, 1});
        return;
      }

      TE->setOperandsInOrder();
      for (unsigned I : seq<unsigned>(0, VL0->getNumOperands()))
        buildTree_rec(TE->getOperand(I), Depth + 1, {TE, I});
```
We need to implement the logics in `VLOperands` and `Tree::setOperandsInOrder`. It will be simpler if we use `VLOperands` only.

Reply:
You can try. But keep in mind, these elements should not be reordered at all. Otherwise, it may affect compile time.

Reply:
I will do this in another PR.

Reply:
see ee74f11629e10ee5b2c9f212b3df62e2f6b70add